src/collector/filter_collector_wrapper.rs

```diff
@@ -168,13 +166,155 @@ where
+/// let top_docs = searcher.search(&query, &filter_collector)?;
+///
+/// assert_eq!(top_docs.len(), 1);
+/// assert_eq!(top_docs[0].1, DocAddress::new(0, 3));
+/// # Ok(())
+/// # }
+/// ```
+pub struct BytesFilterCollector<TCollector, TPredicate>
+where TPredicate: 'static + Clone
+{
+    field: Field,
+    collector: TCollector,
+    predicate: TPredicate,
+}
+
+impl<TCollector, TPredicate> BytesFilterCollector<TCollector, TPredicate>
+where
+    TCollector: Collector + Send + Sync,
+    TPredicate: Fn(&[u8]) -> bool + Send + Sync + Clone,
+{
+    /// Create a new `BytesFilterCollector`.
+    pub fn new(field: Field, predicate: TPredicate, collector: TCollector) -> Self {
+        Self {
+            field,
+            predicate,
+            collector,
+        }
+    }
+}
+
+impl<TCollector, TPredicate> Collector for BytesFilterCollector<TCollector, TPredicate>
+where
+    TCollector: Collector + Send + Sync,
+    TPredicate: 'static + Fn(&[u8]) -> bool + Send + Sync + Clone,
+{
+    type Fruit = TCollector::Fruit;
+
+    type Child = BytesFilterSegmentCollector<TCollector::Child, TPredicate>;
+
+    fn for_segment(
+        &self,
+        segment_local_id: u32,
+        segment_reader: &SegmentReader,
+    ) -> crate::Result<Self::Child> {
+        let schema = segment_reader.schema();
+        let field_name = schema.get_field_name(self.field);
+
+        let column = segment_reader
+            .fast_fields()
+            .bytes(field_name)?
+            .ok_or_else(|| {
+                crate::TantivyError::SchemaError(format!(
+                    "Field `{field_name}` is missing or is not configured as a fast field."
+                ))
+            })?;
+
+        let segment_collector = self
+            .collector
+            .for_segment(segment_local_id, segment_reader)?;
+
+        Ok(BytesFilterSegmentCollector {
+            column,
+            segment_collector,
+            predicate: self.predicate.clone(),
+            buffer: Vec::new(),
+        })
+    }
+
+    fn requires_scoring(&self) -> bool {
+        self.collector.requires_scoring()
+    }
+
+    fn merge_fruits(
+        &self,
+        segment_fruits: Vec<<TCollector::Child as SegmentCollector>::Fruit>,
+    ) -> crate::Result<TCollector::Fruit> {
+        self.collector.merge_fruits(segment_fruits)
+    }
+}
+
+pub struct BytesFilterSegmentCollector<TSegmentCollector, TPredicate>
+where TPredicate: 'static
+{
+    column: BytesColumn,
+    segment_collector: TSegmentCollector,
+    predicate: TPredicate,
+    buffer: Vec<u8>,
+}
+
+impl<TSegmentCollector, TPredicate> SegmentCollector
+    for BytesFilterSegmentCollector<TSegmentCollector, TPredicate>
+where
+    TSegmentCollector: SegmentCollector,
+    TPredicate: 'static + Fn(&[u8]) -> bool + Send + Sync,
+{
+    type Fruit = TSegmentCollector::Fruit;
+
+    fn collect(&mut self, doc: u32, score: Score) {
+        let ord = self.column.ords().values.get_val(doc);
```

Comment:
!?
this line is quite the footgun if you are ddealing with an optional column.

Reply:
> If you have always one value per doc, the mapping is trivial
and your code works.
Otherwise it is plain wrong.

Indeed, one value per doc is my use case. I think this restriction applies to the existing `FilterCollector` as well?

Would you prefer to document this restriction or to try to extend the code to handle multiple values? (I think this would raise the issue if the result of apply the predicate to the values should be combined using "logical or" or "logical and"? Or to pass in the an iterator/vector of values to the predicate?)

Reply:
> Or to pass in the an iterator/vector of values to the predicate?

Noticed that this is somewhat unwieldy without generic closure as the concrete type of the iterator would have to be exposed. So if multiple values should be supported, a boolean flag like `conjunction_of_multiple_values` seems more appropriate. Possibly as constant generic argument `const CONJUNCTION_OF_MULTIPLE_VALUES: bool` to avoid the runtime overhead.

Reply:
I modified the code to at least handle missing values by passing an empty slice to the predicate since this should correspond to passing the default value for the existing `FilterCollector` if I understand things correctly.

Reply:
@fulmicoton I guess I am still doing something wrong in the latest version of the code posted here: I just bumped my project from Tantivy 0.19 to 0.20 and replaced our filter based on `BytesFastFieldReader` to one based on `BytesColumn` as done here and this regressed a single search query from **40ms** to **920ms**. Uff...

I would be grateful for any advice on how to fix the code here, my benchmark or otherwise avoid this.

For example, is it at all possible to avoid the dictionary encoding for high-cardinality fields like mine?

(Basically the same benchmark without accessing the bytes fast field improves from 40ms to 20ms. The code in question is [here](https://gitlab.opencode.de/umwelt-info/metadaten/-/merge_requests/247). I can also share the index itself if there is interest in that, it is just slightly too large to attach here at 43MB.)

Reply:
(With Tantivy 0.19, there version with the filter is basically as fast as the version without it because it does yield fewer documents which completely offsets the extra work done for the filtering. But the cardinalities are exactly the same for 0.19 and 0.20.)

Reply:
((Of course, the filter in question is not the most efficient way to do spatial filtering and we were planning to switch to `rstar` as a spatial index backing this functionality. But with Tantivy 0.19 we would not even need to do that because filtering on a bytes fast field was sufficiently fast.))